### PR TITLE
ActiveRecord: limit() and primary_key

### PR DESCRIPTION
**Pull request by brchristian**
### Summary

Fixes https://github.com/rails/rails/issues/23607.

As discussed in https://github.com/rails/rails/issues/23607, there was an inconsistency between the `first` and `last` ActiveRecord finder methods, in that `limit(1).first` would impose primary key order if no order was specified, whereas `limit(1).last` would not.

Discussion with Rails community members in the issue thread as well as DHH's remarks in https://github.com/rails/rails/pull/23598#issuecomment-189675440 point to the behavior of `last` in this case as being the correct behavior -- that is, that it should not impose primary key order when `limit` has been defined.

This PR includes a test suite for this issue, and simply adds a check for `limit_value.blank?` before imposing primary key ordering by default. This brings the `first` and `last` finder methods into consistent agreement in their behavior on this issue.

### CODE REVIEW

File: activerecord/test/cases/finder_test.rb [hunk 1]
```diff
@@ -1144,6 +1144,23 @@ def test_find_without_primary_key
       Matey.find(1)
     end
   end
+  
+  def test_first_and_last_with_limit_for_order_without_primary_key
+    # While Topic.first should impose an ordering by primary key,
+    # Topic.limit(n).first should not
+    assert_sql(/^((?!ORDER).)*$/) { Post.limit(1).first }
+    assert_sql(/^((?!ORDER).)*$/) { Post.limit(1).last }
```
Comment (sgrif) [inline activerecord/test/cases/finder_test.rb]: We should prefer to test the results of the query, not the generated SQL.

Comment (brchristian) [inline activerecord/test/cases/finder_test.rb]: @sgrif Oh, I was basing it on a similar test `test_find_doesnt_have_implicit_ordering` which I saw in finder_test.rb at https://github.com/rails/rails/blob/master/activerecord/test/cases/finder_test.rb#L307-L309.

Part of the reason I wanted an explicit SQL test like that is that in all databases except Postgres, the default order in which records are returned _is_ ascending by primary key. So the tests based on the results were only failing in Postgres, and only because of how it reacts to the `touch` command (which I would imagine could change in future versions of Postgres). So, I wanted a spec that would fail in all backends, which led me to make the assertion about the SQL itself.

That said, we can remove those SQL assertions and we'll still have a test that fails in current versions of Postgres. If that seems like the better way to go and you'd like me to remove those two SQL tests, let me know and I can add a commit to the PR branch that does that.

**Comment by brchristian** [comment]
@eileencodes Let me know if there is anything I can do to help prepare this PR!

File: activerecord/test/cases/finder_test.rb (hunk 1 above)
Comment (schneems) [inline activerecord/test/cases/finder_test.rb]: Let's remove the sql from these tests you've introduced, keep the other tests that assert SQL in place for this PR. If we want to change it, let's do it in a different PR to keep things focused.

**Comment by schneems** [comment]
Looks like the branch also needs a rebase from master

File: activerecord/test/cases/finder_test.rb (hunk 1 above)
Comment (brchristian) [inline activerecord/test/cases/finder_test.rb]: @schneems Okay, that reasoning makes sense. I’ve added a commit that removes the explicit SQL tests.

Worth noting that with the SQL tests gone, the remaining tests really only apply to Postgres, because SQLite and MySQL impose primary key ordering on their own by default.

**Comment by brchristian** [comment]
Okay, I’ve rebased to the latest `master`, and I also reordered the most recent commit so that the changes to the test suite are all prior to the changes to the code itself, to keep a clean red/green structure.

**Comment by schneems** [comment]
Tests are failing. I’m on mobile now is it related to the change?

**Comment by brchristian** [comment]
Everything’s green except one build that stalled for some reason and timed out. I don’t have permissions to restart it, but maybe you do?

screen shot 2018-07-20 at 7 47 41 pm

**Comment by schneems** [comment]
I restarted it. A trick to getting tests to rebuild is adding white space to the commit message such as an extra newline at the end and then force push. It will start the build again.

**Comment by brchristian** [comment]
Yeah, I was thinking about just pushing the branch again with a different hash, but for some reason I was hesitant to waste the compute that went into the other build jobs by restarting the whole build. But maybe that doesn’t matter in the scheme of things. :) Thanks!

**Comment by bogdanvlviv** [comment]
Should we backport it to `5-2-stable`?

**Comment by schneems** [comment]
Its probably a good idea to have behavior the same in 5.2 and master.

**Comment by kamipo** [comment]
I don't think this should be backported since thisi is a breaking change which changes the existing expectation.

**Comment by kamipo** [comment]
The breaking expectation is the reason not to had been merged by me unlike #27597.
